modules/reactor/src/utils/ReportingIDGeneratorUtils.C

```diff
@@ -0,0 +1,201 @@
+        ++id;
+      }
+      else
+        for (unsigned int k = 0; k < n_cell_elem; ++k)
+          integer_ids.push_back(DofObject::invalid_id);
+    }
+  }
+  return integer_ids;
+}
+
+std::vector<dof_id_type>
+ReportingIDGeneratorUtils::getPatternIntegerIDs(
+    const std::vector<std::unique_ptr<ReplicatedMesh>> & meshes,
+    const std::vector<std::vector<unsigned int>> & pattern)
+{
+  std::vector<dof_id_type> integer_ids;
+  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
+  {
+    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
+    {
+      ReplicatedMesh & cell_mesh = *meshes[pattern[i][j]];
+      unsigned int n_cell_elem = cell_mesh.n_elem();
+      for (unsigned int k = 0; k < n_cell_elem; ++k)
+        integer_ids.push_back(pattern[i][j]);
+    }
+  }
+  return integer_ids;
+}
+
+std::vector<dof_id_type>
+ReportingIDGeneratorUtils::getManualIntegerIDs(
+    const std::vector<std::unique_ptr<ReplicatedMesh>> & meshes,
+    const std::vector<std::vector<unsigned int>> & pattern,
+    const std::vector<std::vector<dof_id_type>> & id_pattern)
+{
+  std::vector<dof_id_type> integer_ids;
+  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
+  {
+    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
+    {
+      unsigned int id = id_pattern[i][j];
+      ReplicatedMesh & cell_mesh = *meshes[pattern[i][j]];
+      unsigned int n_cell_elem = cell_mesh.n_elem();
+      for (unsigned int k = 0; k < n_cell_elem; ++k)
+        integer_ids.push_back(id);
+    }
+  }
+  return integer_ids;
+}
+
+std::set<SubdomainID>
+ReportingIDGeneratorUtils::getCellBlockIDs(
+    const std::vector<std::unique_ptr<ReplicatedMesh>> & meshes,
+    const std::vector<std::vector<unsigned int>> & pattern)
+{
+  std::set<SubdomainID> blks;
+  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
+  {
+    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
+    {
+      ReplicatedMesh & cell_mesh = *meshes[pattern[i][j]];
+      for (auto elem : cell_mesh.element_ptr_range())
+      {
+        auto blk = elem->subdomain_id();
+        auto it = blks.find(blk);
+        if (it == blks.end())
+          blks.insert(blk);
+      }
+    }
+  }
+  return blks;
+}
+
+std::map<SubdomainID, unsigned int>
+ReportingIDGeneratorUtils::getDuckBlockIDs(const std::unique_ptr<MeshBase> & mesh,
+                                           const bool has_assembly_duct,
+                                           const std::set<SubdomainID> & blks)
+{
+  std::map<SubdomainID, unsigned int> blks_duct;
+  if (has_assembly_duct)
+  {
+    unsigned int i = 0;
+    for (auto elem : mesh->element_ptr_range())
+    {
+      auto blk = elem->subdomain_id();
+      auto it1 = blks.find(blk);
+      if (it1 == blks.end())
+      {
+        auto it2 = blks_duct.find(blk);
+        if (it2 == blks_duct.end())
+          blks_duct[blk] = i++;
+      }
+    }
+    blks_duct.erase(blks_duct.begin());
+  }
```

Comment:
```suggestion
  if (has_assembly_duct)
  {
    std::set<SubdomainID> mesh_blks;
    mesh->subdomain_ids(mesh_blks);

    unsigned int i = 0;
    for (const auto mesh_blk : mesh_blks)
      if (!blks.count(mesh_blk))
        blks_duct[blk] = i++;

    // why do you do this? I don't know
    blks_duct.erase(blks_duct.begin());
  }
```

Reply:
```
 blks_duct.erase(blks_duct.begin());
```
The above line is needed to delete the first entry because it is for surrounding regions between the assembly duct and pin (red block in the below figure).

![image](https://user-images.githubusercontent.com/73304529/144368900-89013a4f-b2c8-47d9-b7a0-d54b25d66db6.png)

Comments will be added